geometric_kernels/spaces/graph.py

```diff
@@ -1,27 +1,113 @@
 """
-Graphs: TODO
+Graph object
 """
+
+from typing import Dict, Tuple
+
 import lab as B
+import numpy as np
+import scipy.sparse as sp
 
 from geometric_kernels.eigenfunctions import Eigenfunctions
+from geometric_kernels.lab_extras import degree, eigenpairs, set_value, take_along_axis
 from geometric_kernels.spaces.base import DiscreteSpectrumSpace
 
-# from networkx import Graph
+
+class ConvertEigenvectorsToEigenfunctions(Eigenfunctions):
+    """
+    Converts the array of eigenvectors to a callable objects,
+    where inputs are given by the indices. Based on
+    from geometric_kernels.spaces.mesh import ConvertEigenvectorsToEigenfunctions.
+    TODO(AR): Combine this and mesh.ConvertEigenvectorsToEigenfunctions.
+    """
+
+    def __init__(self, eigenvectors: B.Numeric):
+        """
+        :param eigenvectors: [Nv, M]
+        """
+        self.eigenvectors = eigenvectors
+
+    def __call__(self, X: B.Numeric, **parameters) -> B.Numeric:
+        """
+        Selects `N` locations from the `M` eigenvectors.
+
+        :param X: indices [N, 1]
+        :param parameters: unused
+        :return: [N, M]
+        """
+        indices = B.cast(B.dtype_int(X), X)
+        Phi = take_along_axis(self.eigenvectors, indices, axis=0)
+        return Phi
+
+    def num_eigenfunctions(self) -> int:
+        """Number of eigenvectos, M"""
+        return B.shape(self.eigenvectors)[-1]
 
 
 class Graph(DiscreteSpectrumSpace):
-    """TODO"""
+    """
+    Represents an arbitrary undirected graph.
+    """
+
+    def __init__(self, adjacency_matrix: Tuple[np.array, sp.spmatrix]):  # type: ignore
+        """
+        :param adjacency_matrix: An n-dimensional square, symmetric, binary
+            matrix, where adjacency_matrix[i, j] is one if there is an edge
+            between nodes i and j.
+        """
```

Comment:
maybe add some simple assertions to check that `adjacency_matrix` is indeed square, symmetric and binary? Otherwise, throw an error with an understandable error message for the user.